pcdsdevices/signal.py

```diff
@@ -152,60 +175,251 @@ def _calc_readback(self):
+            )
+
+        # Only check connectivity status of the signal; cross fingers that it
+        # reflects both being connected and having a not-None value.
+        return all(signal.connected for signal in self._signals)
+
+    @contextlib.contextmanager
+    def _check_connectivity(self) -> Generator[Dict[str, bool], None, None]:
+        """
+        Context manager for checking connectivity.
+
+        The block for the context manager should perform some operation
+        to change the state of the signal.
+
+        Returns state information as a dictionary, accessible after the
+        context manager block has finished evaluation.
+        """
+        was_connected = self.connected
+        state_info = {
+            "was_connected": was_connected,
+            "is_connected": was_connected,
+            "sent_md_callback": False,
+            "sent_value_callback": False,
+        }
+
+        yield state_info
+
+        # Now that the caller has updated state, check to see if our
+        # connectivity status has changed.  Update the mutable return
+        # value at this point.
+        is_connected = self.connected
+        state_info["is_connected"] = is_connected
+        if was_connected != is_connected:
+            self._metadata["connected"] = is_connected
+            self._run_metadata_callbacks()
+            state_info["sent_md_callback"] = True
+
+        if not was_connected and is_connected:
+            # disconnected -> connected; give a proper value callback.
+            # "connected" indicates that:
+            # 1. All underlying signals are connected
+            # 2. All underlying signals have a value cached
+            with self._lock:
+                old_value = self._readback
+                self._readback = self._calc_readback()
+                self._run_subs(
+                    sub_type=self.SUB_VALUE,
+                    obj=self,
+                    value=self._readback,
+                    old_value=old_value,
+                )
+                state_info["sent_value_callback"] = True
+
+    def add_signal_by_attr_name(self, name: str) -> Signal:
+        """
+        Add a signal from which to aggregate information.
+
+        This must be called before any subscriptions are made to the signal.
+        Duplicate signals will not be re-added.
+
+        Parameters
+        ----------
+        name : str
+            The attribute name of the signal, relative to the parent device.
+
+        Returns
+        -------
+        sig : ophyd.Signal
+            The signal referenced by the attribute name.
+
+        Raises
+        ------
+        RuntimError
+            If called after .subscribe() or used without a parent Device.
+        """
+        if self._has_subscribed:
+            raise RuntimeError(
+                "Cannot add signals to an AggregateSignal after it has been "
+                "subscribed to."
+            )
+
+        sig = self.parent
```

Comment:
(More musings, less of a suggestion/request).  

I suppose there's no case where we would want to walk further up the component chain to get information from the parent's parent?  

Say we have something with the structure:
```
TopLevelDevice
    component1
        sub1_cpt(MultiDerivedSignal)
        ...
    component2
        sub2_cpt(MultiDerivedSignal)
        ...
```

Then `sub2_cpt` wouldn't be able to reach information about `sub1_cpt`.  It could, however, reach anything within `component2`

I don't necessarily think we should make changes to support this; it might just be easier to flatten the components if we expect them to talk.

Reply:
I suppose you could just specify an attr as `'.parent.component1.sub1_cpt...'`, though that seems subversive

Reply:
I can't yet imagine a scenario where I (personally) would choose to implement a device/component hierarchy like that. That doesn't exclude it from being a possibility, of course. If this pcdsdevices feature sees any use at all outside of just me, I'd expect the question to pop up at some point.

The resolution is - as you reasoned - use `.parent.cpt` to reach your biological great-uncle + aunt components.